### ISSUE TITLE
@inheritdoc should be {@inheritdoc}

### ISSUE THREAD

**Issue by michaelarnauts** [yiisoft/yii2#11635]
According to [the documentation of phpdoc](https://manual.phpdoc.org/HTMLSmartyConverter/HandS/phpDocumentor/tutorial_tags.inlineinheritdoc.pkg.html), the `@inheritdoc` tag to indicate that the function documentation should be copied from the parent function should be written as `{@inheritdoc}`.

Currently, `@inheritdoc` is used everywhere in the code of Yii2.

The next version of Netbeans should be able to process these according to [this page](https://netbeans.org/bugzilla/show_bug.cgi?id=202763).

**Comment by Faryshta** [comment]
i think the use case there is distinct.

```php
/**
 * @inheritdoc
 */
public function method();
```

will copy the long description, and other tags like params, return, etc.

```php
/**
 * {@inheritdoc}
 * @return string
 */
public function method();
```

Will copy ONLY the long description, not the rest of the tags which have to be asigned again.

**Comment by cebe** [comment]
The problem with this is, that there is no real spec of PHPDoc ([yet](https://github.com/php-fig/fig-standards/pull/169)) and at the time when the yii2-apidoc tool has been implemented the page you linked also contained an `@inheritdoc` tag that was described to work similar to what is implemented now. It is actually still there in the PSR proposal: https://github.com/phpDocumentor/fig-standards/blob/b994d853b8ba33bd4bfb00e53cc268d57a5d1e7b/proposed/phpdoc.md#61-making-inheritance-explicit-using-the-inheritdoc-tag so it is hard to say which notation is correct, because our usage of `@inheritdoc` is meant to inherit everything and not just description, it is the correct tag to use.

**Comment by samdark** [comment]
IDEs are OK with what's used currently.

**Comment by rob006** [comment]
> IDEs are OK with what's used currently.

Well, this is not true. I'm testing NetBeans 8.2-RC now and only `{@inheritdoc}` works properly. With `@inheritdoc` it drops method description and actually works like without any `inheritdoc` tag.

![70e4c28d](https://cloud.githubusercontent.com/assets/5972388/18780917/f7b24a04-817f-11e6-9faa-62cbc8476893.png)

![830c9b01](https://cloud.githubusercontent.com/assets/5972388/18780909/f0566d08-817f-11e6-8b9b-ea885cf55a03.png)

![b51df457](https://cloud.githubusercontent.com/assets/5972388/18780926/fd6347a0-817f-11e6-9842-0dd555ee9295.png)

**Comment by MKiselev** [comment]
Для всех методов где есть тег ```@inheritdoc``` просто добавить ```{}```?

**Comment by cebe** [comment]
@MKiselev no, APIdoc tool must be adjusted, see https://github.com/yiisoft/yii2/issues/11635#issuecomment-221995372

**Comment by MKiselev** [comment]
@cebe yep, I was read this PSR.

**Comment by cebe** [comment]
As long as there is no spec to follow I do not think there is anything we can do about this issue.
There is a significant differentce between `@inheritdoc` and `{@inheritdoc}`. `@inheritdoc` will inherit all tags, `{@inheritdoc}` will only inherit description. We can not easily change Yii code to use `{@inheritdoc}`.

**Comment by rob006** [comment]
@cebe "all tags" are inherited by default, so `{@inheritdoc}` and `@inheritdoc` would work exactly the same in 99% cases. And remaining 1% contains mostly situations when `@inheritdoc` is used as inline tag: https://github.com/yiisoft/yii2/blob/master/framework/db/Query.php#L558-L574 . So basically Yii 2 already (incorrectly) uses `@inheritdoc` as inline tag. 

But the main problem with `@inheritdoc` is that nobody (except Yii 2) using it  - everyone sticks to `{@inheritdoc}`. So `@inheritdoc` will always have worse support than `{@inheritdoc}`. We already have great phpdoc with broken links because nobody supports custom `[[]]` syntax, do we really need another Yii-specific syntax in phpdoc?

**Comment by sergeymakinen** [comment]
I tend to agree with @rob006.

```
/**
 * @inheritDoc
 */
```
blocks should be removed at all as they simply don't make any sense.

Only inline `@{inheritDoc}` should be left when they're needed.

**Comment by cebe** [comment]
What exactly is not working with current Yii PHPdoc?

**Comment by rob006** [comment]
Links to classes/methods/properties. Yii 2 uses `[[]]` syntax which is supported only by yii2-apidoc (note that `[[select]]` is not parsed here):

![cfc096a5](https://user-images.githubusercontent.com/5972388/27955107-7ecd1e18-6313-11e7-895a-9ec939965bd0.png)

Yii 1.1 uses standard `{@link}` which works perfectly fine in PhpStorm: 

![7a170fbe](https://user-images.githubusercontent.com/5972388/27955149-b7956d68-6313-11e7-8433-44f481a6bbf5.png)

**Comment by klimov-paul** [comment]
How does usage of `{@link}` is realte to `@inheritdoc` topic?
If you are concern about it - you should open a separated issue.

**Comment by rob006** [comment]
@klimov-paul `{@link}` is only another example of yii-specific syntax in phpdoc (after using `@inheritdoc` instead of `{@inheritdoc}`. But you're right, it should have another issue.

**Comment by bizley** [comment]
I'll just drop the link to [phpDocumentor 2](https://www.phpdoc.org/docs/latest/guides/inheritance.html#the-inheritdoc-tag) mentioning the `@inheritdoc`.

**Comment by sergeymakinen** [comment]
TBH, this issue mostly describes problems with a documentation parsing in IDEs. In a long term, I think it's easier for PhpStorm to support `@inheritDoc`.

**Comment by cebe** [comment]
we're going to make a standard and then follow it ;) https://twitter.com/GeeH/status/884356241430122496

**Comment by rob006** [comment]
Personally I don't see any sense in existing of `@inheritdoc`. It was created only because someone see problem in docblock with `{@inheritdoc}` only. So now we have special non-inline tag which can be used only in one scenario and result is exactly the same as with inline tag. It adds only unnecessary complexity, nothing more.

**Comment by jblac** [comment]
What we've noticed, is phpDocumentor generates errors with `@inheritdoc` but with `{@inheritdoc}` it does not.   Take it for what it's worth -- I've started having my team use the latter option because I don't like my error report being above 0... but, i'm anal that way.

**Comment by klimov-paul** [comment]
See https://github.com/yiisoft/yii2-apidoc/issues/160

**Comment by samdark** [comment]
# Summary of current IDE support

NetBeans looks way too old.

## Block-level `@inheritdoc`

### Eclipse PDT

- `@inheritdoc` doesn't work.

### Netbeans 8.2

- `@inheritdoc` doesn't work.

### PhpStorm

- `@inheritDoc` works the same as `@inheritdoc`.
- If there's only `@inheritdoc`, both description and meta-tags are inherited.
- If there are description and `@inheritdoc`, it overrides description but inherits meta-tags.
- `@since` isn't inherited.
- If there are `@inheritdoc` and meta-tags, it inherits description and adds new meta-tags to the ones from parent class.
  Weird thing is that same-named `@param` tags are duplicated:
  

Base

```
/**     
 * @param int $a
 * @param int $b
 */
```
 
and

```
/**
 * @inheritdoc
 * @param string $b
 * @param int $a
 */
```
 
gives: 

```
string $b
int $a
int $a
int $b
``` 


## Inline `{@inheritdoc}`

### Eclipse PDT

- `{@inheritdoc}` doesn't work.

### NetBeans 8.2

- `{@inheritDoc}` works for classes, `@since` is inherited.
- `{@inheritDoc}` does not work for methods.

### PhpStorm

- `{@inheritDoc}` works like block-level `@inheritdoc` while `{@inheritdoc}` (lowercase) works as expected replacing only
  the token with inehrited description.
- If there's only `{@inheritdoc}`, both description and meta-tags are inherited.
- `@since` isn't inherited.
- If there is description and `{@inheritdoc}` in it, then `{@inheritdoc}` is replaced with inherited description. Meta-tags
  are inherited.
  
- If there are `{@inheritdoc}` and meta-tags, it behaves by replacing `{@inheritdoc}` with description and weird-merging
  meta-tags.

**Comment by samdark** [comment]
That basically means that there's no sense to switch to `{@inheritdoc}`.

**Comment by klimov-paul** [comment]
So we consider @rob006's [screenshots](https://github.com/yiisoft/yii2/issues/11635#issuecomment-249144703) to be forged in Photoshop or something?

Looks like there are 2 researches here with quite an opposite results, so it is hard to determine, who speaks the truth.

**Comment by klimov-paul** [comment]
> it is nice to have since there's PSR for it and it may be used in projects.

PHPDocumentor project uses `{@inheritdoc}` tag with brackets. See:
https://github.com/phpDocumentor/Reflection/blob/develop/src/phpDocumentor/Reflection/Php/Argument.php#L63-L65
https://github.com/phpDocumentor/ReflectionDocBlock/blob/3.3.2/src/DocBlock/Tags/PropertyRead.php#L50-L53

This project is the closest thing to be a standard about PHPDoc at the present state.
We use thier libraries to create our own APIDoc at [yii2-apidoc](https://github.com/yiisoft/yii2-apidoc/blob/master/composer.json#L25-L26).

**Comment by samdark** [comment]
Hmm... let me reinstall NetBeans and try again. I've tried by mouse + CTRL over class name (works) and same over method name (displays `@inheritdoc`).

**Comment by samdark** [comment]
NetBeans 8.2 release:

![tags](https://user-images.githubusercontent.com/47294/35171898-73af0282-fd76-11e7-994e-8f3d54d04472.png)

**Comment by samdark** [comment]
@rob006 
[inheritdoc_playground.zip](https://github.com/yiisoft/yii2/files/1647918/inheritdoc_playground.zip)

**Comment by rob006** [comment]
@samdark `{@inheritdoc}` should go on its own line. Using it  as `this is {@inheritdoc} string` does not make any sense.

**Comment by klimov-paul** [comment]
@samdark, you are mixing different things here.
The example you provide [here](https://github.com/yiisoft/yii2/issues/11635#issuecomment-359090091) is doomed to fail and does not relates to this issue.

PHPDoc has a specific demand on formatting: description plain text goes first, then tags `@param`, `@return` and so on.

In your example to attempt to forge method PHPDoc to be following:

```php
/**
 * This is
 * Parent description
 * @param int $var
 * @return int
 * string
 */
```

which is obviously incorrect.

**Comment by klimov-paul** [comment]
The question here is whether IDE picks up parent description from `@inheritdoc` (plain) as well as from `{@inheritdoc}` (with brackets) correctly or not.

E.g. which has better support:

this?
```php
/**
 * @inheritdoc
 */
public function foo() {..}

/**
 * @inheritdoc
 * @return Override
 */
public function foo() {..}

/**
 * @inheritdoc
 * @param Override $var
 */
public function foo($var) {..}
```
or this?

```php
/**
 * {@inheritdoc}
 */
public function foo() {..}

/**
 * {@inheritdoc}
 * @return Override
 */
public function foo() {..}

/**
 * {@inheritdoc}
 * @param Override $var
 */
public function foo($var) {..}
```

**Comment by samdark** [comment]
OK. Netbeans supports plain `{@inheritdoc}` well. Also checked phpdocumentor code. It doesn't support `@inheritdoc` w/o `{` and `}` at all so yeah, it makes sense to have it.

@rob006, @klimov-paul my example makes sense according to phpdocumenor docs:

- http://docs.phpdoc.org/guides/inheritance.html
- https://manual.phpdoc.org/HTMLSmartyConverter/HandS/phpDocumentor/tutorial_tags.inlineinheritdoc.pkg.html

**Comment by samdark** [comment]
So I'm all for `{@inheritdoc}` w/o any extra strings or property overrides. It has the best support among phpdocumentor, phpstorm and netbeans.

**Comment by klimov-paul** [comment]
OK, since `yii2-apidoc` now supports brackets for 'inheritdoc', I will update the source code in the next few days.

**Comment by cebe** [comment]
Does `{@inheritdoc}` pick up the short or long description or both?

```php
/**
 * This is the short description.
 *
 * This is the long description...
 */
```

**Comment by samdark** [comment]
Both.

**Comment by cebe** [comment]
here is an example, where at the position of `@inheritdoc` only the long description of the parent method should be inserted: https://github.com/yiisoft/yii2/blob/248415422f85dee805ff4cfe3b0a666edc5cdc2a/framework/db/Query.php#L695

**Comment by rob006** [comment]
@cebe There is no real standard for this, so every IDE/tool behaves differently. AFAIK, according to phpDocumentor docs, summary and description inherits separately, so in this `Query` example only summary should be inserted in the place of `{@inheritdoc}`.

BTW: I've made some research some time ago. Not sure if it is still valid, but prepare for brainfuck: https://netbeans.org/bugzilla/show_bug.cgi?id=269107#c8 :D

**Comment by klimov-paul** [comment]
Resolved by commit 65f6b59b35cc63aee7f16a64aeafcec11a8dba39
